dds/idl/be_global.cpp

```diff
@@ -748,12 +757,108 @@ ExtensibilityKind BE_GlobalData::extensibility(AST_Decl* node) const
   return value;
 }
 
+AutoidKind BE_GlobalData::autoid(AST_Decl* node) const
+{
+  AutoidAnnotation* autoid_annotation =
+    dynamic_cast<AutoidAnnotation*>(
+      builtin_annotations_["::@autoid"]);
+  AutoidKind value = autoid_annotation->absent_value;
+  autoid_annotation->node_value_exists(node, value);
+  return value;
+}
+
+bool BE_GlobalData::id(AST_Decl* node, ACE_CDR::ULong& value) const
+{
+  IdAnnotation* id_annotation =
+    dynamic_cast<IdAnnotation*>(builtin_annotations_["::@id"]);
+  return id_annotation->node_value_exists(node, value);
+}
+
+bool BE_GlobalData::hashid(AST_Decl* node, std::string& value) const
+{
+  HashidAnnotation* hashid_annotation =
+    dynamic_cast<HashidAnnotation*>(builtin_annotations_["::@hashid"]);
+  return hashid_annotation->node_value_exists(node, value);
+}
+
+bool BE_GlobalData::is_optional(AST_Decl* node) const
+{
+  OptionalAnnotation* optional_annotation =
+    dynamic_cast<OptionalAnnotation*>(
+      builtin_annotations_["::@optional"]);
+  bool value = optional_annotation->absent_value;
+  optional_annotation->node_value_exists(node, value);
+  return value;
+}
+
+bool BE_GlobalData::is_must_understand(AST_Decl* node) const
+{
+  MustUnderstandAnnotation* must_understand_annotation =
+    dynamic_cast<MustUnderstandAnnotation*>(
+      builtin_annotations_["::@must_understand"]);
+  bool value = must_understand_annotation->absent_value;
+  must_understand_annotation->node_value_exists(node, value);
+  return value;
+}
+
+bool BE_GlobalData::is_key(AST_Decl* node) const
```

Comment:
`BE_GlobalData::check_key`?

Reply:
Didn't want the extra parameter.

Reply:
The extra parameter is significant.

Reply:
There are three possibilities:  (1) the field is annotated with `@key` and a value, (2) the field is annotated with `@key` and no value, or (3) the field does not have the `@key` annotation.  My function returns false for (3) and (1) when the value is false.  This is adequate for my purposes.

Reply:
Okay, I'm still not familiar with type object, but you're talking about `MemberFlag.IS_KEY` just being a boolean option? That is troubling because the way our keys work is a little more complicated than that.

Reply:
No.  What I'm saying is that we cannot assume that a non-topic type is nested.  Again, I don't know why a non-topic non-nested type would be useful, but a user can declare such a thing.  Thus, if code was checking for nested by looking at !topic, then we need to go change the code to actually check for nested.

Reply:
@iguessthislldo Please resolve this conversation or let me know what other information you need.